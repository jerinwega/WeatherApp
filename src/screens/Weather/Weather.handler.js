import { useEffect, useState, useCallback, useContext } from 'react';
import { Keyboard } from 'react-native';
import { useDispatch, useSelector } from 'react-redux';
import { debounce, get } from 'lodash';
import { fetchWeatherByCity } from '../../redux/actions/weatherAction';
import { getData, storeData } from '@/src/utils/asyncStorage';
import { ThemeContext } from '../../../App';
import { WHITE_COLOR, DARK_THEME, LIGHT_THEME, YELLOW_400 } from '../../utils/constants';


const WeatherHandler = () => {
  const dispatch = useDispatch();
  const { colorMode } = useContext(ThemeContext);

  const { weather, isFetching, error } = useSelector(state => state.weather);

  const [loading, setLoading] = useState(false);
  const [showSearch, toggleSearch] = useState(false);
  const [canShowWeather, setCanShowWeather] = useState(true);

  useEffect(() => {
    fetchInitialWeatherByCity();
  }, []);

  const fetchInitialWeatherByCity = async () => {
    const savedCity = await getData('city');
    if (savedCity) {
      dispatch(fetchWeatherByCity(savedCity));
    }
  };

  const handleSearch = async (value) => {
    setLoading(true);
    setCanShowWeather(false);

    if (value.length > 2) {
      try {
        const result = await dispatch(fetchWeatherByCity(value));

        if (get(result, 'city')) {
          storeData('city', get(result, 'city'));
          setLoading(true); // NOTE: maybe should be false?
        } else {
          setLoading(false);
        }
      } catch (error) {
        setLoading(false);
        console.error('Search error:', error);
      }
    }
  };

  const handleToggleSearch = () => {
    setLoading(false);
    Keyboard.dismiss();
    toggleSearch(!showSearch);

    if (get(weather, 'city')) {
      setCanShowWeather(true);
    }
  };

  const handleSearchDebounce = useCallback(debounce(handleSearch, 200), []);

  const gradientColors =
    colorMode === 'dark'
      ? [DARK_THEME, YELLOW_400] // black to yellow-400
      : [WHITE_COLOR, LIGHT_THEME]; // white to blue-600

  return {
    colorMode,
    loading,
    setLoading,
    showSearch,
    toggleSearch,
    canShowWeather,
    setCanShowWeather,
    handleToggleSearch,
    handleSearchDebounce,
    gradientColors,
    weather,
    isFetching,
    error,
  };
};

export default WeatherHandler;
