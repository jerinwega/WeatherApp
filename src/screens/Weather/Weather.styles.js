import { Dimensions, StyleSheet } from 'react-native';
import { WHITE_COLOR } from '../../utils/constants'; // adjust path if needed

const { height: deviceHeight } = Dimensions.get('window');
const { width: deviceWidth } = Dimensions.get('window');

const styles = StyleSheet.create({
    background: {
    position: 'absolute',
    top: 0,
    left: 0,
    right: 0,
    height: deviceHeight,
    width: deviceWidth,
    zIndex: -1, // ensures it's behind everything
  },
  searchContainer: {
    borderRadius: 9999,
    overflow: 'hidden',
  },
  searchContainerVisible: {
    backgroundColor: WHITE_COLOR,
  },
  searchContainerHidden: {
    backgroundColor: 'transparent',
  },
});

export default styles;
