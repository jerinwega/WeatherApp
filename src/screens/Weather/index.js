import React from 'react';
import { TouchableOpacity, View, TextInput, Keyboard } from 'react-native';
import styles from './Weather.styles';
import { Text } from "@/src/components/ui"
import { Image as CachedImage } from 'expo-image';
import Header from '../../components/Header/Header';
import { get } from 'lodash';
import { LinearGradient } from 'expo-linear-gradient';
import { Search } from 'lucide-react-native';
import * as Progress from 'react-native-progress';
import { Dimensions } from 'react-native';
import { WHITE_COLOR, BLACK_COLOR, DARK_THEME, LIGHT_THEME } from '../../utils/constants';
import WeatherHandler from './Weather.handler';

const Weather = () => {

const { width } = Dimensions.get('window');
const imageSize = width * 0.55; // 55% of screen width

const {
    colorMode,
    loading,
    showSearch,
    handleToggleSearch,
    handleSearchDebounce,
    canShowWeather,
    weather,
    isFetching,
    error,
    gradientColors,
  } = WeatherHandler();

let foreCastFrag = null;

if (!isFetching && error) {
  foreCastFrag = (
  <View className='flex-1 flex-row justify-center items-center'>
    <Text size='xl' bold className='text-red-800 bg-red-200 px-6 py-4 rounded-3xl'>{error}</Text>
  </View> 
  );
} else if (canShowWeather && get(weather, 'city')) {
foreCastFrag = (
<>
{/* forecast section */}
<View className='flex justify-center flex-1'>
    <Text className={`text-center ${colorMode === 'dark' ? 'text-white': 'text-black'}`} size='3xl' bold>{get(weather, 'city')}, 
      <Text className={`text-center ${colorMode === 'dark' ? 'text-white': 'text-black'}`} size='xl' bold>
        &nbsp;{get(weather, 'country')}
      </Text>
      </Text>
  <View className='flex-row flex justify-center'>
    <CachedImage
        style={{ width: imageSize, height: imageSize }}
        source={{
          uri: get(weather, 'icon'),
        }}
        contentFit="contain"
      />              
    </View>
    <View className='space-y-2'>
          <Text className={`text-center ml-5 ${colorMode === 'dark' ? 'text-white': 'text-black'}`} size='6xl' bold>{get(weather, 'temperature')}&#176;</Text>
          <Text className={`text-center ${colorMode === 'dark' ? 'text-white' : 'text-black'}`} size='xl' bold>{get(weather, 'condition')}</Text>
    </View>
  </View> 
  </>
);
}

return (
  <>
  <View className='flex-1 relative'>
    <LinearGradient
      colors={gradientColors}
      style={styles.background}
      start={colorMode === 'dark' && {x: 0, y: 0}}
      end={colorMode === 'dark' && {x: 0, y: 12}}
    />

    <View className='z-50 p-3'>
      <Header />
    </View>

{/* searchbar section */}
<View className='m-5 z-50 shadow-xl' style={{ height: '7%' }}>
  <View className='flex-row justify-end items-center rounded-full' style={{ backgroundColor: showSearch ? 'transparent' : WHITE_COLOR }}>
  {showSearch ? null : 
  (
    <TextInput
      onChangeText={handleSearchDebounce}
      placeholder='Search city'
      placeholderTextColor="black"
    className="flex-1 text-base font-semibold text-black px-6"
    style={{
      lineHeight: 16
    }}
    />
  )
  }
  <TouchableOpacity
  onPress={handleToggleSearch}
  style={{ backgroundColor: LIGHT_THEME}}
  className='rounded-full p-3 m-1'
  >
  <Search size={24} color={colorMode === 'dark'? WHITE_COLOR : BLACK_COLOR} />

  </TouchableOpacity>
  </View>
</View>

  {loading ? (
    <View className='flex-1 flex-row justify-center items-center'>
    <Progress.CircleSnail thickness={10} size={140} color={colorMode=== 'dark' ? LIGHT_THEME : DARK_THEME} />
    </View>
  ) : 
  foreCastFrag 
  } 
  </View> 
  </>
  )
};

export default Weather;

