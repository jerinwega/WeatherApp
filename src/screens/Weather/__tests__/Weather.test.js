import React from 'react';
import { render } from '@testing-library/react-native';
import Weather from '../index';

// Mock expo-image
jest.mock('expo-image', () => ({
  Image: () => null,
}));

// Mock expo-linear-gradient
jest.mock('expo-linear-gradient', () => {
  const { View } = require('react-native');
  return {
    LinearGradient: View,
  };
});

// Mock lucide-react-native
jest.mock('lucide-react-native', () => ({
  Search: () => null,
}));

// Mock react-native-progress
jest.mock('react-native-progress', () => ({
  CircleSnail: () => null,
}));

// Mock Header component
jest.mock('../../../components/Header/Header', () => () => <></>);

// Mock Weather.handler.js
jest.mock('../Weather.handler', () => () => ({
  colorMode: 'light',
  loading: false,
  showSearch: false,
  handleToggleSearch: jest.fn(),
  handleSearchDebounce: jest.fn(),
  canShowWeather: true,
  weather: {
    city: 'London',
    country: 'UK',
    icon: 'https://example.com/icon.png',
    temperature: 22,
    condition: 'Partly Cloudy',
  },
  isFetching: false,
  error: null,
  gradientColors: ['#ffffff', '#0000ff'],
}));

describe('Weather Screen', () => {
  it('renders correctly and matches snapshot', () => {
    const { toJSON } = render(<Weather />);
    expect(toJSON()).toMatchSnapshot();
  });
});
