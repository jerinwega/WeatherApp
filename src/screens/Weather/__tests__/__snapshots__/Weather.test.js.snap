// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Weather Screen renders correctly and matches snapshot 1`] = `
<View
  className="flex-1 relative"
>
  <View
    colors={
      [
        "#ffffff",
        "#0000ff",
      ]
    }
    end={false}
    start={false}
    style={
      {
        "height": 1334,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "width": 750,
        "zIndex": -1,
      }
    }
  />
  <View
    className="z-50 p-3"
  />
  <View
    className="m-5 z-50 shadow-xl"
    style={
      {
        "height": "7%",
      }
    }
  >
    <View
      className="flex-row justify-end items-center rounded-full"
      style={
        {
          "backgroundColor": "#FFFFFF",
        }
      }
    >
      <TextInput
        className="flex-1 text-base font-semibold text-black px-6"
        onChangeText={[MockFunction]}
        placeholder="Search city"
        placeholderTextColor="black"
        style={
          {
            "lineHeight": 16,
          }
        }
      />
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "backgroundColor": "#2563eb",
            "opacity": 1,
          }
        }
      />
    </View>
  </View>
  <View
    className="flex justify-center flex-1"
  >
    <Text
      className="font-body font-bold text-3xl text-center text-black"
    >
      London
      ,
      <Text
        className="font-body font-bold text-xl text-center text-black"
      >
         
        UK
      </Text>
    </Text>
    <View
      className="flex-row flex justify-center"
    />
    <View
      className="space-y-2"
    >
      <Text
        className="font-body font-bold text-6xl text-center ml-5 text-black"
      >
        22
        °
      </Text>
      <Text
        className="font-body font-bold text-xl text-center text-black"
      >
        Partly Cloudy
      </Text>
    </View>
  </View>
</View>
`;
